Sources/NIOCore/ByteBuffer-core.swift

```diff
@@ -1119,3 +1117,100 @@ extension ByteBuffer {
         return Range<Int>(uncheckedBounds: (lower: indexFromReaderIndex, upper: upperBound))
     }
 }
+
+// MARK: HexDump helpers
+
+extension ByteBuffer {
+
+    /// Return a `String` of hexadecimal digits of bytes in the Buffer,
+    /// in a format that's compatible with `xxd -r -p`.
+    ///
+    /// - parameters:
+    ///     - offset:  The offset from the beginning of the buffer, in bytes. Defaults to 0.
+    ///     - readableOnly: Whether to only dump the `readableBytes` part (before the `writerIndex`), or to dump the full buffer, including empty bytes in the end of the allocated space. Defaults to true.
+    func hexDumpShort(offset: Int = 0, readableOnly: Bool = true) -> String {
+        let length = (readableOnly ? self.writerIndex : self.capacity) - offset
+        return self._storage.dumpBytes(slice: self._storage.fullSlice,
```

Comment:
- would you mind moving the byte dumping out of `ByteBuffer-core` into `ByteBuffer-aux` as it's not "core" functionality
- the hex dumping functionality should use `ByteBuffer`'s _public_ API. I.e. only things that can be accessed from the outside and especially no `self._*` things

Reply:
Honestly, I would delete (or leave alone) the old `self._storage.dumpBytes()`. That was written to debug the underlying storage itself when we're replacing the storage. So essentially it's an internal debugging facility that we can delete.

This new hex dumping facility probably need no APIs but

- `self.withUnsafeReadableBytes { ... }` <- pointer to all readable bytes
- `self.readableBytesView` <- a `Collection` of all readable bytes
- `self.getInteger(...)` <- a third way to access the bytes
- `self.readerIndex`/`self.writerIndex`/`self.redableBytes` <- how many/which bytes are readable

all of them are public.

Reply:
> Honestly, I would delete (or leave alone) the old self._storage.dumpBytes().
Sure. Since hex dump functions will cover similar functionality (but at a higher level), I'll delete the old `dumpBytes` from Storage and clean up the places where we use it to make sure i don't break anything.

> Even better: make a new file for it ByteBuffer+HexDumps.swift
EZ. I thought of moving them, and just started in the `core` file to quickly prototype. 

> he hex dumping functionality should use ByteBuffer's public API. I.e. only things that can be accessed from the outside and especially no self._* things

Yep, got it. 

> self.withUnsafeReadableBytes { ... } <- pointer to all readable bytes

This will be the easiest rewrite from what's currently in the pull request. 

At first, I started out using `readableBytes`, but I used it in a way that would introduce a bug, and I only read a single byte with `getInteger` because I didn't know if reading the whole buffer would make a copy of the data (and hence did not understand how memory efficient that would be).

Let me go clean this up.

Reply:
> Honestly, I would delete (or leave alone) the old `self._storage.dumpBytes()`. That was written to debug the underlying storage itself when we're replacing the storage. So essentially it's an internal debugging facility that we can delete.
> 
> This new hex dumping facility probably need no APIs but
> 
> * `self.withUnsafeReadableBytes { ... }` <- pointer to all readable bytes
> * `self.readableBytesView` <- a `Collection` of all readable bytes
> * `self.getInteger(...)` <- a third way to access the bytes
> * `self.readerIndex`/`self.writerIndex`/`self.redableBytes` <- how many/which bytes are readable
> 
> all of them are public.

Another implicit API which might be of use is just grabbing a mutable copy of the buffer as the backing storage will only copy-on-write (CoW) if you modify the _underlying data_. This may simplify the length limited cases:

```swift
let originalBuffer = ...
// buffer and originalBuffer share underlying storage for data but do not share
// their indexes
var buffer = originalBuffer
// moves readerIndex on buffer but not originalBuffer
let prefix = buffer.readSlice(length: 64)
let prefixHex = prefix.hexDump()

buffer.advanceReaderIndex(by: ...)
let suffixHex = buffer.hexDump()
```